sentry_streams/sentry_streams/adapters/arroyo/rust_step.py

```diff
@@ -273,23 +273,56 @@ class ArroyoStrategyDelegate(RustOperatorDelegate, Generic[TStrategyIn, TStrateg
     def __init__(
         self,
         inner: ProcessingStrategy[TStrategyIn],
-        in_transformer: Callable[[RustMessage, Committable], ArroyoMessage[TStrategyIn]],
+        in_transformer: Callable[[KafkaMessage, Committable], ArroyoMessage[TStrategyIn]],
         retriever: OutputRetriever[TStrategyOut],
     ) -> None:
         self.__inner = inner
         self.__in_transformer = in_transformer
         self.__retriever = retriever
+        self.__watermarks: MutableSequence[PyWatermark] = []
+        # globbed_committable is the combined committable of all messages polled
+        # from the OutputRetriever
+        self.__globbed_committable: Committable = {}
+
+    def __should_send_watermark(self, watermark: PyWatermark) -> bool:
+        """
+        Returns True if:
+        - all partitions in `watermark`'s payload are present in `self.__globbed_committable`
+        - all offsets for those partitions in `watermark` are less than or equal to the corresponding
+        offsets in `self.__globbed_committable`
+        """
+        try:
+            for partition, offset in watermark.payload.items():
+                if self.__globbed_committable[partition] < offset:
+                    return False
+        except KeyError:
+            return False
+        return True
+
+    def __yield_messages(self) -> Iterable[Tuple[RustMessage, Committable]]:
+        for message, committable in self.__retriever.fetch():
+            yield (message, committable)
+            self.__globbed_committable.update(committable)
+            watermarks = self.__watermarks
+            for wm in watermarks:
+                if self.__should_send_watermark(wm):
+                    yield (wm, wm.payload)
+                    self.__watermarks.remove(wm)
```

Comment:
This logic will be quite delicate. It is very important to have it thoroughly tested. Now we are relying on the multiprocess delegate and reduce delegate tests to cover this. Such approach is fragile because those tests may assume the ArroyoStrategyDelegate works and not cover this logic. 

There is a `test_delegate.py`  file that tests the `RustOperatorDelegate` in a simpler environment than multiprocess and Reduce. Please add test cases there to validate that:
- The delegate sends through the right Watermark after seeing the committables
- The corner cases where partial committables are returned so we need to accumulate multiple of them before sending a watermark 
- Whether the watermark is send before or after the message in the right way.